lxd/container_lxc.go

```diff
@@ -5973,6 +6015,135 @@ func (c *containerLXC) createNetworkDevice(name string, m types.Device) (string,
+	}
+
+	newDevice := types.Device{}
+	err := shared.DeepCopy(&m, &newDevice)
+	if err != nil {
+		return nil, err
+	}
+
+	if !shared.PathExists(fmt.Sprintf("/sys/class/net/%s", m["parent"])) {
+		return nil, fmt.Errorf("Parent device '%s' doesn't exist", m["parent"])
+	}
+	sriovNumVFs := fmt.Sprintf("/sys/class/net/%s/device/sriov_numvfs", m["parent"])
+	sriovTotalVFs := fmt.Sprintf("/sys/class/net/%s/device/sriov_totalvfs", m["parent"])
+
+	// verify that this is indeed a SR-IOV enabled device
+	if !shared.PathExists(sriovTotalVFs) {
+		return nil, fmt.Errorf("Parent device '%s' doesn't support SR-IOV", m["parent"])
+	}
+
+	// get number of currently enabled VFs
+	sriovNumVfsBuf, err := ioutil.ReadFile(sriovNumVFs)
+	if err != nil {
+		return nil, err
+	}
+	sriovNumVfsStr := strings.TrimSpace(string(sriovNumVfsBuf))
+	sriovNum, err := strconv.Atoi(sriovNumVfsStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// get number of possible VFs
+	sriovTotalVfsBuf, err := ioutil.ReadFile(sriovTotalVFs)
+	if err != nil {
+		return nil, err
+	}
+	sriovTotalVfsStr := strings.TrimSpace(string(sriovTotalVfsBuf))
+	sriovTotal, err := strconv.Atoi(sriovTotalVfsStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if any VFs are already enabled
+	vf := ""
+	nicName := ""
+	for i := 0; i < sriovNum; i++ {
+		vf = fmt.Sprintf("virtfn%d", i)
+		if !shared.PathExists(fmt.Sprintf("/sys/class/net/%s/device/%s/net", m["parent"], vf)) {
+			vf = ""
+			continue
+		}
+
+		// Check if VF is already in use
+		empty, err := shared.PathIsEmpty(fmt.Sprintf("/sys/class/net/%s/device/%s/net", m["parent"], vf))
+		if err != nil {
+			return nil, err
+		}
+		if empty {
+			vf = ""
+			continue
+		}
+
+		vf = fmt.Sprintf("/sys/class/net/%s/device/%s/net", m["parent"], vf)
+		ents, err := ioutil.ReadDir(vf)
+		if err != nil {
+			return nil, err
+		}
+		if len(ents) == 0 || len(ents) > 1 {
+			continue
+		}
+
+		if shared.StringInSlice(ents[0].Name(), reserved) {
+			continue
+		}
+
+		nicName = ents[0].Name()
+
+		// found free VF
+		break
+	}
+
+	if nicName == "" {
+		if sriovNum == sriovTotal {
+			return nil, fmt.Errorf("All virtual functions of vfio device '%s' seem to be in use", m["parent"])
+		}
+
+		// bump the number of VFs to the maximum
+		err := ioutil.WriteFile(sriovNumVFs, []byte(sriovTotalVfsStr), 0644)
+		if err != nil {
+			return nil, err
+		}
+
+		// use next free VF index
+		vf = fmt.Sprintf("virtfn%d", sriovNum+1)
+
+		for i := sriovNum + 1; i < sriovTotal; i++ {
+			vf = fmt.Sprintf("/sys/class/net/%s/device/%s/net", m["parent"], vf)
+			ents, err := ioutil.ReadDir(vf)
+			if err != nil {
+				return nil, err
+			}
+			if len(ents) == 0 || len(ents) > 1 {
+				return nil, fmt.Errorf("Failed to determine unique device name")
+			}
+			if shared.StringInSlice(ents[0].Name(), reserved) {
+				continue
+			}
+
+			nicName = ents[0].Name()
+		}
+	}
+
+	if nicName == "" {
+		return nil, fmt.Errorf("All virtual function allocated")
```

Comment:
"All virtual functions are already used on device %s" would probably be more useful.

Reply:
done